Migration of QuestionObjectFactory (Redo)

h-ARTS opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

~~1. This PR fixes or fixes part of #[fill_in_number_here].~~
2. This PR does the following: Migration of QuestionObjectFactory 

## Essential Checklist

- ~~[ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)~~
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

abhayraghuwanshi committed:
added docstrings (#9028)

rohitkatlaa committed:
Fixes #5254 Shuffle multiple choice (#8738)

* shuffle multiple choice answers

* draft upgrade services

* requested changes

* merge confilct

* changes

* added tests to question services

* changes to the order of tests

* removing addition yaml content

* requested changes

* linting

* draft upgrade service

* requested changes

* requested changes

* changes in draft

* requested changes

* requested changes

* requested changes

* use css instead of by repeater

* draft true to false

* failing e2e test

nadintamer committed:
Fix #8758: add waitFor statement before accessing response tab element (#9081)

* Add waitFor.visibilityOf before accessing response tab after switching page

* Fix typo

* Test response tab flake broken version

vojtechjelinek committed:
Fix #9106: Upgrade batch of libraries, fix frontend tests, and other small changes  (#9107)

* Remove unused files

* Use toEqual instead of toBe to make the failed cases easier to debug

* Update build source to more reflect current state of the codebase

* Clean imports and requires

* Put constants into the previous state after test are run

* Fix typo

* Fix frontend tests

* Upgrade libraries

* Fix import

* Revert core-js upgrade

* Revert upgrade of html-loader

vojtechjelinek committed:
Fix #9014, #9016: Add empty callback function when modal is cancelled (#9116)

aks681 committed:
Changed character limits for the new editors and allowed saving of invalid exps in draft stories (#9087)

* Changed character limits and allowing saving of stories with invalid exps when not public

* fixed frontend tests

* minor change

* concept card modal changes

* made review changes

* fixed backend errors

* increased backend coverage

* made review changes

* made review changes

* made review changes

marianadasilvadev committed:
Increase 100% coverage files that were fakely fully covered. (#9096)

* Increase to 100% files that were fakely fully covered

* Remove one spec from  QuestionObjectFactorySpec since #9094 is covering it

* Fix review comments

vojtechjelinek committed:
Update authors and changelog for v2.9.2 (#9128)

* Update CHANGELOG

* Update AUTHORS

* Update CONTRIBUTORS

* Update core/templates/pages/about-page/about-page.directive.html

* Fix Jim Zhan

U8NWXD committed:
Fix #9086 and #9100: Fix classroom e2e flakes (#9118)

* Wait for Topic Thumbnail Visibility Before Editing

Before editing the topic thumbnail, wait for it to be visible.

Fixes #9086.

* Remove Duplicate Create Topic Button HTML Tag

When no topics have been created, the topics and skills dashboard page
shows a create topic button in addition to the button in the navigation
bar. These two buttons shared the `protractor-test-create-topic-button`
HTML tag, which caused Protractor to raise warnings. This commit removes
the HTML tag from the button outside the navigation bar.

* First Wait for Role Update Confirmation Visibility

We were waiting for the role update confirmation message to first
contain the expected text and then to be visible. This commit swaps the
order of those checks in an attempt to address #9100 and #9097.

safwan20 committed:
Fix part of #4057: Added a test file for  exploration-param-changes.service.ts (#9114)

* test-1

* test-2

* test-3

* test-4

abhayraghuwanshi committed:
Fix part of #4057. Unit test for exploration-feature-service.ts (#9076)

* made service file

* the exploration data object changed

* service test added

* added to 100% coverage

* sorted alphabtically

* changed the exploration feature tests

* requested change done

* fixed the linter tests

bansalnitish committed:
Fix part of #8472: Upgrades CollectionCreationBackendApiService to Angular (#9125)

* Upgrades CollectionCreationBackendApiService to Angular

* addresses comments

DubeySandeep committed:
Add a lint check for all jobs register in jobs_registry.py file (#9129)

* Add test for job registry

* Fix removal issue.

* Removed unwanted try-catch

* Address review comments.

* Removed test for non-existing one-off job

* Removed unwanted newline!

* Address review comments (2)

* Undo unwanted changes

* Removed unwanted tests.

aks681 committed:
Fix #9121: Address review comments in #9087 (#9126)

* fix #9121

* made review  changes

* review change

aks681 committed:
Fix UI and UX issues in new structures editors and dashboard (#9110)

* Fix issues in new structures editors

* e2e test

* minor changes

* fix cramped skill cards

* fixed e2e test: added tab management

* rename value

* made review changes

* made review change

tjinjoy committed:
Fixes issue #3369: Resize editor to avoid overflowing when creator uses longer responses (#9122)

* Resize contenteditable textbox to prevent response overflow #3369

* Changed spacing in ckeditor directive

* Fix errors from lint and typescript tests

* Fix quoted width property lint error

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into develop

h-ARTS committed:
Migrate QuestionObjectFactory

bansalnitish reviewed (approved):
Thanks @h-ARTS, LGTM!

iamprayush commented:
@h-ARTS Please note the failing CI tests.

h-ARTS commented:
> @h-ARTS Please note the failing CI tests.

@iamprayush I'm still waiting for @bansalnitish response on gitter/hangouts. He's already aware of this. In the meantime, I'm working on `thread-data.service.ts`

bansalnitish commented:
Sincere apologies for the delay @h-ARTS, I was busy this complete week. I'll definitely work on this travis for you today.

bansalnitish commented:
I checked the `creatorDashboard.js` suite and I just ran `'should display correct stats on dashboard'`. It is failing with this error:

![image](https://user-images.githubusercontent.com/22434689/82113277-76989b00-9772-11ea-9f55-73b7abb720c9.png)

I added `CamelCaseHyphensPipe` to the `exploration-player-page-module.ts` given the error but it's still failing. Also, I'm not able to reproduce this locally. Could you please help me out @ankita240796 @aks681?

Also, it is not a good practice to add all the pipes in each module. We will create a `shared-component.module.ts` if this one gets solved. Refer to this answer: https://stackoverflow.com/a/42577318

Thanks!

h-ARTS commented:
@ankita240796 @aks681 Ping!

h-ARTS committed:
Merge branch 'develop' into question-object-f

h-ARTS committed:
Wrong version number passed in a test case

h-ARTS commented:
@bansalnitish Can you rerun accessibility, file_upload_features, and users e2e ci's? Timeouts..

bansalnitish commented:
These tests were passing before you merged develop, right? 

(Could you please merge it again may be -- I don't think there is should be any issue with this PR since tests passed before)

h-ARTS commented:
> These tests were passing before you merged develop, right?
> 
> (Could you please merge it again may be -- I don't think there is should be any issue with this PR since tests passed before)

@bansalnitish That's correct but how would I do this?

bansalnitish commented:
Do this: 

1. git checkout develop
2. git pull upstream develop
3. git checkout branch-name
4. git merge develop

Let me know if you need help. Thanks!

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into question-object-f

iamprayush commented:
@h-ARTS Please note the failing CI tests, thanks!

h-ARTS commented:
> @h-ARTS Please note the failing CI tests, thanks!

@iamprayush Thanks but I'm still waiting for @ankita240796 and @aks681 response. I don't know how to proceed with this because they are flaking long ago since my old PR's!

kevinlee12 commented:
@bansalnitish can you advise? thanks!

ankita240796 commented:
> > @h-ARTS Please note the failing CI tests, thanks!
> 
> @iamprayush Thanks but I'm still waiting for @ankita240796 and @aks681 response. I don't know how to proceed with this because they are flaking long ago since my old PR's!

@h-ARTS can you let us know what help do you need specifically here?

bansalnitish commented:
@ankita240796 could you please refer to this comment
: https://github.com/oppia/oppia/pull/9286#issuecomment-629601281?

ankita240796 commented:
> I checked the `creatorDashboard.js` suite and I just ran `'should display correct stats on dashboard'`. It is failing with this error:
> 
> ![image](https://user-images.githubusercontent.com/22434689/82113277-76989b00-9772-11ea-9f55-73b7abb720c9.png)
> 
> I added `CamelCaseHyphensPipe` to the `exploration-player-page-module.ts` given the error but it's still failing. Also, I'm not able to reproduce this locally. Could you please help me out @ankita240796 @aks681?
> 
> Also, it is not a good practice to add all the pipes in each module. We will create a `shared-component.module.ts` if this one gets solved. Refer to this answer: https://stackoverflow.com/a/42577318
> 
> Thanks!

Hi @bansalnitish, I am still not sure what you want me to do here exactly. Can you please describe what are the things you want me to do. Thanks!

aks681 reviewed (changes requested):
@bansalnitish Do you mean this error?
![image](https://user-images.githubusercontent.com/22347970/83258939-cfd1e700-a1d4-11ea-86f6-6c603d8b1147.png)

After creating a question in develop, in the skill editor, if I refresh the page after switching to this branch (and updating this branch with develop), I get this error.
@h-ARTS Try updating your branch with develop and trying it out locally?

bansalnitish commented:
@h-ARTS could you please refer to the above comment and let us know what happens?

h-ARTS commented:
@aks681 @bansalnitish I updated my branch with developing and tried out locally. Here are the commits that I see locally since I updated:
```
*   c64a4e5 Merge remote-tracking branch 'upstream/develop' into question-object-f
|\  
| * cef57a3 Fix #8472: Collection creation service upgrade (#9303)
| * a340ef4 Fixes #5436 Randomize pretest questions (#9061)
| * 33c068d Fixed user updates. (#9252)
| * f1c0f0a Update ExplorationMissingStatsAudit to read directly from raw models (#9384)
:...skipping...
*   c64a4e5 Merge remote-tracking branch 'upstream/develop' into question-object-f
|\  
| * cef57a3 Fix #8472: Collection creation service upgrade (#9303)
| * a340ef4 Fixes #5436 Randomize pretest questions (#9061)
| * 33c068d Fixed user updates. (#9252)
| * f1c0f0a Update ExplorationMissingStatsAudit to read directly from raw models (#9384)
| * a4a3a72 Fix part of #6271: Modify add_misconception to accept object instead of dict (#9351)
| * b6ba041 Fix part of #7176: Add type definitions in some object factories (#9345)
| * a1fa806 Remove any from some object factories (#9370)
| * 4b1a0f5 Fix part of #8472: Upgrades UserExplorationPermissionsService to Angular (#9338)
| * 204a287 Removes codeowners for community dashboard pages. (#9369)
| * 7439c1f Add WARNING_TYPE to interaction validation spec files (#9371)
| * 4e041be Allow saving of stories with explorations containing skillreview RTE and DragAndDropSortInput interaction (#9360)
| * fac762c Routine upgrade of libs (#9359)
| * 48809be Fix part of #7176: Remove any types from some object factories - II (#9349)
| * 9cfb5e9 Extend ExplorationMissingStatsAudit to check for state stats (#9361)
| * 4a59747 Creation of .lighthouserc.js file and scripts/lighthouse_checks.py file (#9295)
| * 3b89e1f Migrates Background Banner directive to Angular Component (#9330)
| * 3ed6b09 Transfer codeowners (#9355)
| * 1d468ac Fix part of #8472: Upgrades CurrentInteractionService to Angular (#9347)
| * 56b24a5 Fix part of #8472: Upgrades ExplorationRecommendationsService to Angular (#9339)
| * 0b9d90a Fixes part of #8668: All the Frontend services should be documented with jsdoc (#9250)
| * a9b1e47 Fix #8464: Wraps exceptions as an Error object if it does not have a message property (#9343)
| * b397ce0 Milestone 1.5: Add a check for checking the version of third party type defs (#9310)
| * 4627a0a Exception handling is used to handle unexpected errors (#9273)
| * c2706b5 Removing all nonmigratable instances of $scope and $rootScope from skill-editor module (#9136)
| * 1a1d9d9 Fix #7493: Remove jinja from base.py (#9294)
| * 39fff8e Adding @marianazangrossi as codeowner from all frontend spec files. (#9323)
| * 1d499bc Fix #9331: Added typescript checks to pre push hook (#9334)
| * d93e1f2  Added update username functionality in the admin tab. (#9191)
| * 7d15d81 Update authors and changelog for v2.9.3 (#9332)
| * b95f0e9 Fix part of #8472: Migration (topics-and-skills-dashboard-backend-api.service.ts) AngularJs to Angular 8 (#8856)
* |   56e3af6 Merge remote-tracking branch 'upstream/develop' into question-object-f
|\ \  
| |/  
| * 7c61186 Fix #9309: Added conditional while rendering the classroom page. (#9312)
| * b9ef6e6 Fix part of #8472: Upgrade StateInteractionStatsService to Angular 8 (#9321)
* | 032ea4e Wrong version number passed in a test case
* |   2b20ca1 Merge branch 'develop' into question-object-f
|\ \  
Hanans-MacBook-Pro:oppia hananmufti$ git log --graph --since=2.weeks
*   commit c64a4e51e9cbf6a7650155040deced4d99637026
|\  Merge: 56e3af6 cef57a3
| | Author: h-ARTS <hi5hanan@yahoo.com>
| | Date:   Fri May 29 15:41:58 2020 +0200
| | 
| |     Merge remote-tracking branch 'upstream/develop' into question-object-f
| |   
| * commit cef57a36abf5b72fe03401cd14521fd8a44517c6
| | Author: Hadyn Fitzgerald <me@hadyn.dev>
| | Date:   Fri May 29 06:47:35 2020 -0600
| | 
| |     Fix #8472: Collection creation service upgrade (#9303)
| |     
| |     * Upgraded collection creation service.
| |     
| |     Migrated the collection creation service to Angular from AngularJS. Added unit tests for collection creation service that did not exist before.
| |     
| |     * Fixed linter errors
| |     
| |     * Added service to 100% coverage whitelist
| |     
| |     * Fixed TypeScript tests
| |     
| |     * Refactored Collection Creation service
| |     * Moved using Location service to using window reference.
| |         * Updated tests to reflect this change.
| |     * Added service to upgraded services typescript.
| |     
| |     * Merged
| |     
| |     * Fixed issue injector issue
| |     
| |     * Made recommended changes
| |     
| |     * Fixed test dependency on AlertsService
| |     
| |     * Fixed linting issue
| |     
| |     * Removed CollectionCreationService from test coverage blacklist
| |   
| * commit a340ef4707eb6e76ec77992aa82a06dbf4b07694
```
I tried to create a question but it didn't work as I get the same error as above. Also tried e2e test using only `--suite=skillEditor` and it didn't work either:

<img width="1421" alt="Screenshot 2020-05-29 at 17 04 34" src="https://user-images.githubusercontent.com/11369764/83274767-7ff02180-a1ce-11ea-9abf-7c9b1097717f.png">

bansalnitish commented:
I see there is something happening with CamelCaseToHyphensPipe not being found. I'll look into this more and get back here within a day. 

(I think adding CamelCaseToHyphensPipe in shared-module will do the trick -- will check this)

bansalnitish commented:
Hi @h-ARTS, One potential fix for this would be to add CamelCaseToHyphensPipe in the shared-module.ts (We have shared-module now and we will add all pipes there, let's begin with CamelCase and see if it works). I have not yet checked it locally by pulling your branch.
